### ISSUE TITLE
problem with keyboard input and ^C

### ISSUE THREAD

**Issue by francoisvignon** [tbodt/ish#246]
it was a semantic issue with ^C management (I propose a correction at the end of this note):

if you key in following:

sleep 3600[enter]
echo abc[enter]

and before sleep is terminated you press ^C, sleep is aborted, fine, but all pending input before made before ^C is pressed are processed, and the result is performing the next action after the one aborted:
```
iFv:~# sleep 3600                                
echo abc                                         
^C                                               
iFv:~# echo abc                                  
abc        
```

so, maybe another example:
you start a task (task1) and you anticipate the end by keying next task (task2) and you see something is wrong during task1, you press ^C and then task2 is started.

I know, you can key in:  task1 ; task2 <enter> and then ^C abort all, but it's another discussion ;-)

another view of the problem: you start a task, this one will ask something after sometime, you anticipate the answer and before the task ask you, you press ^C, then the answer not used by the task was still in the input processing.

a last view of this: you past some lines of script, and if you press ^C, only the current line is aborted, next one was performed (or at least, started) until you press ^C (in fact, you should press ^C for each jobs)

pasting following
```
sleep 3600
echo a
sleep 3600
echo b
sleep 3600
echo c
```
to stop this one, you should press ^C three times

**the semantic of ^C should be: flush input buffer and send SIGINT to current process**

**Comment by tbodt** [comment]
I checked and this is the correct behavior, unless the NOFLSH flag is set in c_lflag of the termios.

**Comment by francoisvignon** [comment]
So, maybe it's the correct behavior, but it's not the behavior of MacOs, Debian, and most Unix system I know :-S

In bash (or other) copy/past my last example (with the three sleep), and the first ^C you hit kill the sequence, not only the first sleep.

so, maybe the wrongness come from something not defining correctly NOFLSH flag

**Comment by francoisvignon** [comment]
more informations and one question:
settings tty without or without noflsh don't affect comportement:

```
Fv:~# stty                                      
speed 0 baud; line = 0;                          
intr = ^C; quit = ^\; erase = ^?; kill = ^U;     
eof = ^D; eol = <undef>; eol2 = <undef>;         
swtch = <undef>; start = ^Q; stop = ^S;          
susp = ^Z; rprnt = ^R; werase = ^W; lnext = ^V;  
flush = ^O; min = 1; time = 0;                   
-cread                                           
-brkint -imaxbel                                 
-iexten -echok -echoke        
```

stty noflsh
```
Fv:~# stty                                      
speed 0 baud; line = 0;                          
intr = ^C; quit = ^\; erase = ^?; kill = ^U;     
eof = ^D; eol = <undef>; eol2 = <undef>;         
swtch = <undef>; start = ^Q; stop = ^S;          
susp = ^Z; rprnt = ^R; werase = ^W; lnext = ^V;  
flush = ^O; min = 1; time = 0;                   
-cread                                           
-brkint -imaxbel                                 
-iexten -echok noflsh -echoke
```

stty -noflsh
```
Fv:~# stty                                      
speed 0 baud; line = 0;                          
intr = ^C; quit = ^\; erase = ^?; kill = ^U;     
eof = ^D; eol = <undef>; eol2 = <undef>;         
swtch = <undef>; start = ^Q; stop = ^S;          
susp = ^Z; rprnt = ^R; werase = ^W; lnext = ^V;  
flush = ^O; min = 1; time = 0;                   
-cread                                           
-brkint -imaxbel                                 
-iexten -echok -echoke        
```

with both settings, input flush is not performed with ^C.

maybe it's a misconversion of NOFLSH termio flags ? I see value of this one differ from MacOs/iOs/BSD and Linux ...

but I can't find any occurence of NOFLSH in iSH. do you use it to clear or not input buffer in your tty driver/emulator when receiving a ^C ?

**Comment by tbodt** [comment]
Until now I didn't know that ^C was supposed to flush the input buffer. So I'm going to add code to flush the buffer unless NOFLSH is set. I just tried setting NOFLSH in a bash prompt on Mac and got this:

```
bash-3.2$ stty noflsh
bash-3.2$ sleep 100
echo asdf
^C
bash-3.2$ echo asdf
asdf
bash-3.2$
```

**Comment by francoisvignon** [comment]
thanks to have try :-)

and yes with NOFLSH set, it's correct  and by default, NOFLSH is not set.

thx for your time.

oh, a missing thing: Happy new year for you and your close :-)


send from my mobile.
____________________

François

**Comment by francoisvignon** [comment]
this issue was (from my point of view) solved by commit  9ed9bc93ecc44db6b3bea9611c01912fe6eb6ad9 (and so, can be closed ?)

many thanks @tbodt

**Comment by tbodt** [comment]
oh, I forgot to close.
